### ISSUE TITLE
[Ascend] Random sampling (temperature > 0) always returns token 0 — output degenerates to !!!!

### ISSUE THREAD

**Issue by Al2O3Al2O3** [flagos-ai/vllm-plugin-FL#542]
# [Ascend] Random sampling (temperature > 0) always returns token 0 — output degenerates to `!!!!`

**Repo:** https://github.com/flagos-ai/FlagTree/ (FL / vllm_fl ascend integration)

## Summary

On the Ascend backend, **any request with `temperature > 0` degenerates**: the
model emits token id 0 every step, so the completion is a run of `!` characters.
Greedy decoding (`temperature = 0`, pure `argmax`) produces correct output. The
failure is independent of top-k/top-p — it also happens with `temperature` alone
(no `top_k`/`top_p`), and even at `temperature = 0.01` (which should behave
almost like greedy).

This is a **separate defect** from the Triton `_topk_topp_kernel` compile crash
(that one only triggers at batch ≥ 8 via the Triton path). This one is on the
PyTorch fallback path (batch < 8, or with a forced fallback), which is the path
one would normally use to work around the compile crash — so the two together
leave no working sampling path on Ascend.

## Environment

Same image as the companion compile-crash issue:
flagtree `0.7.0rc2+ascend3.5`, vllm `0.20.2+flagos`, flag_gems
`5.4.0rc2.post3+gad89ec477`, torch/torch_npu `2.10.0`, CANN `9.0.0`,
target `Ascend910_9382` (910C), TP=2.

## Reproduction

Serve on Ascend (TP=2, `fl` plugin), send single (batch=1) requests:

```
# greedy — CORRECT
prompt="北京是" temperature=0 max_tokens=16
→ "中国的政治中心，也是中国的文化中心。..."

# any sampling — BROKEN (token 0 repeated)
prompt="北京是" temperature=1.0                          → "!!!!!!!!!!!!!!!!"
prompt="北京是" temperature=0.7 top_p=0.95               → "!!!!!!!!!!!!!!!!"
prompt="北京是" temperature=0.7 top_k=20                 → "!!!!!!!!!!!!!!!!"
prompt="北京是" temperature=0.01                         → "!!!!!!!!!!!!!!!!"
```

## Investigation done

Each sampling sub-op was unit-tested **offline on the NPU** and behaves
correctly in isolation:

- `torch.Tensor.exponential_()` — produces a valid exponential distribution
  (values ~2e-4 .. ~12), not all-zero.
- `apply_top_k_top_p_pytorch(logits, k, p)` at real vocab size (151936) —
  softmax sums to 1, no NaN, ~18 non-zero entries for `top_k=20/top_p=0.95`,
  diverse argmax.
- `random_sample(probs, {})` with empty generators at vocab 151936 — draws a
  diverse, valid token (not 0).

So every component passes offline, yet online random sampling always yields
token 0. The most likely root cause is in `random_sample`
(`vllm/v1/sample/ops/topk_topp_sampler.py`):

```python
q = torch.empty_like(probs)
if len(generators) != probs.shape[0]:
    q.exponential_()
...
return probs.div_(q).argmax(dim=-1).view(-1)
```

Hypothesis: in the live inference stream on NPU, `q` from `torch.empty_like`
is not reliably filled by `exponential_()` (async execution / uninitialized
memory not synchronized), so `probs.div_(q)` yields inf/NaN and `argmax`
collapses to index 0. Offline repros pass because standalone calls carry an
implicit sync. **Not yet confirmed by in-process instrumentation** — needs a
hook inside the running sampler to capture `q` / `probs.div_(q)` at request time.

## Impact

No usable sampling on Ascend. Combined with the compile-crash issue:

| Path | Condition | Result |
|---|---|---|
| Triton top-k/top-p | batch ≥ 8 | engine crash (compile) |
| PyTorch fallback | batch < 8, temperature > 0 | wrong output (token 0) |
| Greedy | temperature = 0 | correct |

## Ask

Confirm and fix random sampling on the Ascend backend. If the cause is the
`empty_like`+`exponential_` pattern not being NPU-safe, an explicit fill /
synchronization (or a `torch.rand`-based Gumbel path) should be considered for
the Ascend integration.

**Comment by Al2O3Al2O3** [comment]
Root cause has been pinned down and it is NOT in vllm-plugin-FL. In-server instrumentation showed the probs tensor is correct (sums to 1, no NaN) but q from Tensor.exponential_() contains 1024 NaNs, which collapses argmax to token 0. Minimal repro confirms FlagGems exponential_ (flag_gems/ops/exponential_.py) emits exactly 1024 NaNs for tensors with n >= 65536 on Ascend, while uniform_() on the same tensor is fine. Moving this to FlagGems: flagos-ai/FlagGems#6446. Closing here.
